components/nexus-security/src/main/java/org/apache/shiro/nexus/NexusSessionFactory.java

```diff
@@ -29,9 +33,49 @@
 {
   private static final Logger log = LoggerFactory.getLogger(NexusSessionFactory.class);
 
+  /**
+   * Ensure {@link SimpleSessionImpl} is used and provides logging.
+   */
   @Override
   public Session createSession(final SessionContext initData) {
     log.trace("Creating session w/init-data: {}", initData);
-    return super.createSession(initData);
+
+    // duplicated from SimpleSessionFactory, retaining class-hierarchy for sanity
+    if (initData != null) {
+      String host = initData.getHost();
+      if (host != null) {
+        return new SimpleSessionImpl(host);
+      }
+    }
+    return new SimpleSessionImpl();
+  }
+
+  /**
+   * Customized session impl to apply synchronized-treatment to attributes map.
+   */
+  private static class SimpleSessionImpl
+    extends SimpleSession
+  {
+    public SimpleSessionImpl() {
+      super();
+    }
+
+    public SimpleSessionImpl(final String host) {
+      super(host);
+    }
+
+    /**
+     * Work around bug in Shiro which uses a non-synchronized map to back attributes.
+     *
+     * This appears to only be called by {@link SimpleSession#getAttributesLazy()}.
+     */
+    @Override
+    public void setAttributes(final Map<Object, Object> attributes) {
+      // NOTE: usage seems to be non-null, but guard for sanity
+      if (attributes != null) {
+        super.setAttributes(Collections.synchronizedMap(attributes));
+      }
+      super.setAttributes(null);
```

Comment:
This doesn't look right - you'll end up always setting the attributes to null.

Could just use `super.setAttributes(attributes != null ? Collections.synchronizedMap(attributes) : null);`